pkg/engine/deployment.go

```diff
@@ -181,12 +183,23 @@ func newDeployment(ctx *Context, info *deploymentContext, opts deploymentOptions
 	if err != nil {
 		return nil, err
 	}
-	plugctx = plugctx.WithCancelChannel(ctx.Cancel.Canceled())
+
+	// Keep the plugin context open until the context is terminated, to allow for graceful provider cancellation.
+	plugctx = plugctx.WithCancelChannel(ctx.Cancel.Terminated())
+
+	// Set up a goroutine that will signal cancellation to the source if the caller context
+	// is cancelled.
+	cancelCtx, cancelFunc := context.WithCancel(context.Background())
+	go func() {
+		<-ctx.Cancel.Canceled()
+		logging.V(7).Infof("engine.newDeployment(...): received cancellation signal")
+		cancelFunc()
+	}()
 
 	opts.trustDependencies = proj.TrustResourceDependencies()
 	// Now create the state source.  This may issue an error if it can't create the source.  This entails,
 	// for example, loading any plugins which will be required to execute a program, among other things.
-	source, err := opts.SourceFunc(ctx.BackendClient, opts, proj, pwd, main, projinfo.Root, target, plugctx, dryRun)
+	source, err := opts.SourceFunc(cancelCtx, ctx.BackendClient, opts, proj, pwd, main, projinfo.Root, target, plugctx, dryRun)
```

Comment:
A cancelation context is provided to the source func, and is cancelled eagerly, to abort the plugin installation procedure within the source func.  Prior to this PR, the source func would rely on a context based on `plugctx.Request()` (which is canceled when the plugctx is closed, and is too late for our purposes).